src/core/ddsi/src/ddsi_raweth.c

```diff
@@ -391,6 +395,334 @@ static int joinleave_asm_mcgroup (ddsrt_socket_t socket, int join, const ddsi_lo
+  dds_return_t rc = DDS_RETCODE_OK;
+  ssize_t ret;
+  struct ddsi_vlan_header vhdr;
+  uint16_t vtag;
+  size_t hdrlen;
+  (void) flags;
+
+  assert(msgfrags->niov <= INT_MAX - 1); // we'll be adding one later on
+ 
+  vtag = (uint16_t)(((dst->port >> 20) & 0xfff) + (((dst->port >> 16) & 0xe) << 12));
+  
+  memcpy(vhdr.e.dmac, dst->address + 10, 6);
+  memcpy(vhdr.e.smac, uc->m_base.m_base.gv->interfaces[0].loc.address + 10, 6);
+  
+  if (vtag) {
+    vhdr.e.proto = htons ((uint16_t) ETHERTYPE_VLAN);
+    vhdr.vtag = htons (vtag);
+    vhdr.proto = htons ((uint16_t) uc->m_base.m_base.m_port);
+    hdrlen = sizeof(vhdr);
+  } else {
+    vhdr.e.proto = htons ((uint16_t) uc->m_base.m_base.m_port);
+    hdrlen = 14;
+  }
+
+  DDSRT_STATIC_ASSERT(DDSI_TRAN_RESERVED_IOV_SLOTS >= 1);
+
+  ddsrt_iovec_t * const iovs = (ddsrt_iovec_t *) &msgfrags->tran_reserved[DDSI_TRAN_RESERVED_IOV_SLOTS - 1];
+  iovs[0].iov_base = &vhdr;
+  iovs[0].iov_len = hdrlen;
+
+  if ((ret = writev (uc->m_sock, iovs, (int)(msgfrags->niov + 1))) < 0)
+  {
+      DDS_CERROR(&conn->m_base.gv->logconfig, "ddsi_raweth_conn_write failed with retcode %"PRIdSIZE, ret);
+      rc = DDS_RETCODE_ERROR;
+  }
+
+  return (rc == DDS_RETCODE_OK ? ret : -1);
+}
+
+static dds_return_t ddsi_raweth_set_filter (struct ddsi_tran_factory * fact, ddsrt_socket_t sock, uint32_t port)
+{
+  int r;
+  ushort etype = (ushort)(port & 0xFFFF);
+  struct bpf_insn insns[] = {
+    BPF_STMT(BPF_LD+BPF_H+BPF_ABS, 12),
+	  BPF_JUMP(BPF_JMP+BPF_JEQ+BPF_K, etype, 3, 0),
+    BPF_JUMP(BPF_JMP+BPF_JEQ+BPF_K, ETHERTYPE_VLAN, 0, 3),
+    BPF_STMT(BPF_LD+BPF_H+BPF_ABS, 16),
+    BPF_JUMP(BPF_JMP+BPF_JEQ+BPF_K, etype, 0, 1),
+    BPF_STMT(BPF_RET+BPF_K, (u_int)-1),
+    BPF_STMT(BPF_RET+BPF_K, 0),
+  };
+  unsigned flen = sizeof (insns)/sizeof (struct bpf_insn);
+  struct bpf_program filter = {flen, insns};
+
+  if ((r = ioctl (sock, BIOCSETF, &filter)) == -1 ) {
+    ddsrt_close (sock);
+    DDS_CERROR (&fact->gv->logconfig, "ddsrt_setsockopt attach filter for protocol %u failed ... retcode = %d\n", port, r);
+    return DDS_RETCODE_ERROR;
+  }
+  return DDS_RETCODE_OK;
+}
+
+static dds_return_t ddsi_raweth_create_conn (struct ddsi_tran_conn **conn_out, struct ddsi_tran_factory * fact, uint32_t port, const struct ddsi_tran_qos *qos)
+{
+  int r;
+  int i;
+  dds_return_t rc;
+  ddsrt_socket_t sock = -1;
+  ddsi_raweth_conn_t uc = NULL;
+  struct sockaddr_dl addr = {0};
+  bool mcast = (qos->m_purpose == DDSI_TRAN_QOS_RECV_MC);
+  struct ddsi_domaingv const * const gv = fact->gv;
+  struct ddsi_network_interface const * const intf = qos->m_interface ? qos->m_interface : &gv->interfaces[0];
+  uint32_t buflen;
+
+  if (port == 0 || port > 65535)
+  {
+    DDS_CERROR (&fact->gv->logconfig, "ddsi_raweth_create_conn %s port %u - using port number as ethernet type, %u won't do\n", mcast ? "multicast" : "unicast", port, port);
+    return DDS_RETCODE_ERROR;
+  }
+
+#if defined(DDSI_BPF_IS_CONING_DEV)
```

Comment:
If my suspicion on this above is correct, then this should also be `DDSI_BPF_IS_CLONING_DEV`.